lib/core/src/Cardano/Pool/DB/Sqlite.hs

```diff
@@ -347,6 +355,20 @@ newDBLayer trace fp timeInterpreter = do
                 , Desc PoolRegistrationSlotInternalIndex
                 ]
 
+        , listRetiredPools = \epochNo -> do
+            let query = T.unwords
+                    [ "SELECT * FROM "
+                    , databaseViewName effectivePoolRetirements
+                    , "WHERE retirement_epoch <="
+                    , T.toText epochNo
+                    , ";"
+                    ]
+            let safeCast (Single poolId, Single retirementEpoch) =
+                    PoolRetirementCertificate
+                        <$> fromPersistValue poolId
+                        <*> fromPersistValue retirementEpoch
```

Comment:
Hmmm. Why is the unsafe query needed here? We should stick with the persistent DSL as much as possible and only resort to unsafe queries when the DSL is too limiting or becomes a real performance impediment.

Reply:
@KtorZ wrote:
> Why is the unsafe query needed here?

When determining the current set of effective retirements, we need to combine results from **two different tables**:

1. `pool_registrations`
2. `pool_retirements`

This is because registrations and retirements can _cancel each other out_.

The only way to find out whether a retirement certificate is still in effect is to verify that there are no other retirement certificates _or_ registration certificates that cancel it out.

> we should only resort to unsafe queries when the DSL is too limiting

I do agree with this.

The implementation in this PR determines the set of currently-effective retirement certificates with a query that uses:
- the SQL `UNION` function;
- the SQL `row_number` window function.

As far as I know, `persistent` supports neither of the above functions (at least out of the box).

Furthermore, defining this query in terms of an SQL **view** can help us if we need to debug a real pool database. By opening the database and fetching the `effective_pool_retirements` view, we can immediately see which retirement certificates the wallet currently considers to be effective.

Reply:
Here's the `effective_pool_retirements` view on the pools DB for mainnet:

![effective_pool_retirements](https://user-images.githubusercontent.com/206319/90015613-6c3b5e00-dcdb-11ea-9fd9-0881c41f6e5e.png)

To implement `listRetiredPools`, we just need to select all rows with a `retirement_epoch` that is earlier than or equal to the given epoch.